frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/SlideInBackgroundable.java

```diff
@@ -0,0 +1,166 @@
+package com.virtuslab.gitmachete.frontend.actions.backgroundables;
+
+import static com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle.getString;
+import static com.virtuslab.gitmachete.frontend.vfsutils.GitVfsUtils.getMacheteFilePath;
+import static io.vavr.API.$;
+import static io.vavr.API.Case;
+import static io.vavr.API.Match;
+import static io.vavr.Predicates.instanceOf;
+import static java.text.MessageFormat.format;
+
+import java.nio.file.Path;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vcs.VcsNotifier;
+import git4idea.GitLocalBranch;
+import git4idea.repo.GitRepository;
+import io.vavr.collection.List;
+import io.vavr.control.Try;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import com.virtuslab.branchlayout.api.BranchLayoutEntry;
+import com.virtuslab.branchlayout.api.EntryDoesNotExistException;
+import com.virtuslab.branchlayout.api.EntryIsDescendantOfException;
+import com.virtuslab.branchlayout.api.EntryIsRootException;
+import com.virtuslab.branchlayout.api.IBranchLayout;
+import com.virtuslab.branchlayout.api.IBranchLayoutEntry;
+import com.virtuslab.branchlayout.api.readwrite.IBranchLayoutWriter;
+import com.virtuslab.gitmachete.frontend.actions.dialogs.SlideInOptions;
+
+public class SlideInBackgroundable extends Task.Backgroundable {
+
+  private final GitRepository gitRepository;
+  private final IBranchLayout branchLayout;
+  private final IBranchLayoutWriter branchLayoutWriter;
+  private final Runnable preSlideInRunnable;
+  private final SlideInOptions slideInOptions;
+  private final String parentName;
+  private final VcsNotifier notifier;
+
+  public SlideInBackgroundable(
+      Project project,
+      GitRepository gitRepository,
+      IBranchLayout branchLayout,
+      IBranchLayoutWriter branchLayoutWriter,
+      Runnable preSlideInRunnable,
+      SlideInOptions slideInOptions,
+      String parentName) {
+    super(project, getString("action.GitMachete.BaseSlideInBranchBelowAction.task-title"));
+    this.preSlideInRunnable = preSlideInRunnable;
+    this.gitRepository = gitRepository;
+    this.branchLayout = branchLayout;
+    this.slideInOptions = slideInOptions;
+    this.parentName = parentName;
+    this.notifier = VcsNotifier.getInstance(project);
+    this.branchLayoutWriter = branchLayoutWriter;
+  }
+
+  @Override
+  public void run(ProgressIndicator indicator) {
+    preSlideInRunnable.run();
+
+    waitForLocalBranch();
```

Comment:
IIUC in `preSlideInRunnable` only `FetchBackgroundable` is nonblockin operation.
So, if we create something like `BaseFetch` or just `Fetch` that will **not** be `Task.Backgroundable` (so be blocking fetch operation) then we can probably get ride of `waitForLocalBranch` (maybe use only some kind of simple check if branch exists)

Reply:
Hmm @mkondratek  this indicates that an extra docs is needed why we actually need to wait here... and which of the git4idea operations are *sneakily-asynchronous* (i.e. asynchronous without clearly indicating that fact by e.g. returning a `Future`) :/

Reply:
I didn't think of that but it was only b/c Mikołaj told me in person; so your doubts @micpiotrowski are justified: even though there are more non-blocking/async operations (like `checkoutRemoteBranch` IIRC), it's not clear from just looking at the `BaseSlideInBranchBelowAction`/`SlideInBackgroundable` code